### PR TITLE
task level advanced scheduling policy

### PR DESCRIPTION
**Pull request by hwdef**
ref: #1627

There is generally no problem, but there is currently a bug that has not been resolved.
When the vcjob is deleted, a new vcjob with the same name is created without waiting for all pods to be deleted. The pod in the vcjob will be lost and only a part of it will be created.

### CODE REVIEW

**Comment by hwdef** [comment]
/hold

Cancel the hold until the bug is resolved

**Comment by Thor-wl** [comment]
Suggest to add UT and e2e for this feature first.

File: pkg/controllers/job/helpers/helpers.go [hunk 1]
```diff
@@ -35,7 +36,7 @@ const (
 )
 
 // GetTaskIndex returns task Index.
-func GetTaskIndex(pod *v1.Pod) string {
+func GetPodIndex(pod *v1.Pod) string {
```
Comment (Thor-wl) [inline pkg/controllers/job/helpers/helpers.go:39]: Suggest to reserve the original name for the uniformity of the name

File: pkg/controllers/job/helpers/helpers.go [hunk 2]
```diff
@@ -70,3 +71,24 @@ func GenPVCName(jobName string) string {
 func GetJobKeyByReq(req *apis.Request) string {
 	return fmt.Sprintf("%s/%s", req.Namespace, req.JobName)
 }
+
+func GetTaskIndex(taskName string, job *batch.Job) int {
```
Comment (Thor-wl) [inline pkg/controllers/job/helpers/helpers.go:75]: Pls add comments for functions as format. Same to the whole PR.

**Review by shinytang6 — CHANGES_REQUESTED** [review]
Will this pr contains changes about scheduler ?

File: pkg/controllers/job/helpers/helpers.go [hunk 3]
```diff
@@ -70,3 +71,24 @@ func GenPVCName(jobName string) string {
...
+}
+
+func GetPodsInTask(taskName string, job *batch.Job) []string {
+	var res []string
+	for _, task := range job.Spec.Tasks {
+		if task.Name == taskName {
+			for index := 0; index < int(task.Replicas); index++ {
+				res = append(res, MakePodName(job.Name, taskName, index))
+			}
+		}
```
Comment (Thor-wl) [inline pkg/controllers/job/helpers/helpers.go:91]: Sugeest to add `break`  to optimize the performance when `task` is found

File: pkg/controllers/job/job_controller_actions.go [hunk 4]
```diff
@@ -307,7 +307,7 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
 	var running, pending, terminating, succeeded, failed, unknown int32
 	taskStatusCount := make(map[string]batch.TaskState)
 
-	var podToCreate []*v1.Pod
+	podToCreate := make(map[string][]*v1.Pod) // taskname -> []pod
```
Comment (Thor-wl) [inline pkg/controllers/job/job_controller_actions.go:310]: Remove nonstandard comments

File: pkg/controllers/job/job_controller_actions.go [hunk 5]
```diff
@@ -349,33 +353,67 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
-					pod.Name, job.Name, err)
-				appendError(&creationErrs, fmt.Errorf("failed to create pod %s, err: %#v", pod.Name, err))
-			} else {
-				classifyAndAddUpPodBaseOnPhase(newPod, &pending, &running, &succeeded, &failed, &unknown)
-				calcPodStatus(pod, taskStatusCount)
-				klog.V(3).Infof("Created Task <%s> of Job <%s/%s>",
-					pod.Name, job.Namespace, job.Name)
+	var dependsOn batch.DependsOn
+
+	for taskName, podToCreateEachTask := range podToCreate {
```
Comment (Thor-wl) [inline pkg/controllers/job/job_controller_actions.go:364]: Suggest to make some optimization about this piece of code in performance.

File: pkg/controllers/job/job_controller_actions.go [hunk 6]
```diff
@@ -349,33 +353,67 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+		go func(taskName string, podToCreateEachTask []*v1.Pod) {
+			taskIndex := jobhelpers.GetTaskIndex(taskName, job)
+			if job.Spec.Tasks[taskIndex].DependsOn != nil {
+				dependsOn = *job.Spec.Tasks[taskIndex].DependsOn
+				if len(job.Spec.Tasks[taskIndex].DependsOn.Name) > 1 && job.Spec.Tasks[taskIndex].DependsOn.Iteration == batch.IterationAny {
+					breakFlag := false
+					for {
+						time.Sleep(time.Second)
```
Comment (Thor-wl) [inline pkg/controllers/job/job_controller_actions.go:372]: Why sleep 1 second?

File: pkg/controllers/job/job_controller_actions.go [hunk 7]
```diff
@@ -349,33 +353,67 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+							}
+						}
+						if breakFlag {
+							break
+						}
+					}
+				} else {
+					for _, DependsOnTask := range dependsOn.Name {
+						for {
+							time.Sleep(time.Second)
```
Comment (Thor-wl) [inline pkg/controllers/job/job_controller_actions.go:386]: Same as above

File: pkg/controllers/job/helpers/helpers.go (hunk 1 above)
Comment (k82cn) [inline pkg/controllers/job/helpers/helpers.go:39]: Let's use `Task` and `TaskInstance` :)

**Comment by hwdef** [comment]
> Will this pr contains changes about scheduler ?

No, it will not change the scheduler.

File: pkg/controllers/job/job_controller_actions.go (hunk 6 above)
Comment (hwdef) [reply pkg/controllers/job/job_controller_actions.go:372]: I think there is no need to judge too fast here.

File: pkg/controllers/job/helpers/helpers.go (hunk 1 above)
Comment (hwdef) [reply pkg/controllers/job/helpers/helpers.go:39]: @Thor-wl 

I think the function of this function is to find the index of the pod in the task, so maybe the name `GetPodIndex` is more appropriate?

Comment (Thor-wl) [reply pkg/controllers/job/helpers/helpers.go:39]: Indeed, it's an awkward point that `Task` in the code represents `Pod` most through the project. It is easily confused for new comers to make difference.
Do you have some good ideas?
@william-wang @shinytang6 @jasonliu747

File: pkg/controllers/job/job_controller_actions.go [hunk 8]
```diff
@@ -349,33 +353,67 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
-			} else {
-				classifyAndAddUpPodBaseOnPhase(newPod, &pending, &running, &succeeded, &failed, &unknown)
-				calcPodStatus(pod, taskStatusCount)
-				klog.V(3).Infof("Created Task <%s> of Job <%s/%s>",
-					pod.Name, job.Namespace, job.Name)
+	var dependsOn batch.DependsOn
+
+	for taskName, podToCreateEachTask := range podToCreate {
+		go func(taskName string, podToCreateEachTask []*v1.Pod) {
+			taskIndex := jobhelpers.GetTaskIndex(taskName, job)
```
Comment (shinytang6) [inline pkg/controllers/job/job_controller_actions.go:366]: suggest to make this block into separate function

**Review by shinytang6 — COMMENTED** [review]
Another tip: it's better to give an obvious reminder that gang need to close if this feature enabled

File: pkg/controllers/job/job_controller_actions.go (hunk 8 above)
Comment (hwdef) [reply pkg/controllers/job/job_controller_actions.go:365]: How about we write this in webhook?

File: pkg/controllers/job/helpers/helpers.go (hunk 1 above)
Comment (william-wang) [reply pkg/controllers/job/helpers/helpers.go:39]: Seems "GetTaskIndex" is the best for two places:)  Let's use GetTask(pod *v1.Pod) here and use the GetTaskIndex(taskName, job) for your new code change.

Comment (hwdef) [reply pkg/controllers/job/helpers/helpers.go:39]: @william-wang ok, that loosk good.

**Comment by hwdef** [comment]
@Thor-wl @shinytang6 @william-wang 

Most of the review comments have been resolved, and the following questions remain:

* What is the appropriate setting for the detection interval? It is currently one nanosecond, this time may be too short https://github.com/volcano-sh/volcano/pull/1833/files#diff-28557d59525e85335ee0eacff9269dcc0dbc0ffcc7ef146ef4e0e59d713df788R473
* Prompt need to disable gang plugin.Here we have two options, one is to limit the minavailable value to the  reasonable, let the Gang plugin can be retained, and the other is to disable the Gang plugin. No matter which kind of choice, we are best completed in Webhook.
* We still have a bug here, as I demonstrated at the weekly meeting. Create the same name Job when Job is not completely deleted. After many attempts, I think this bug is caused by this PR, did not exist before. But I haven't found the reason for this bug yet.Maybe fix this bug later, first merge PR, so we can not block other PR,#1874

File: pkg/controllers/job/job_controller_actions.go [hunk 9]
```diff
@@ -452,6 +468,78 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+						return true, nil
+					}
+					return false, nil
+				})
+			}
+		}
+	}
+}
+
+func (cc *jobcontroller) isDenpendentPodsReady(DependsOnTask string, job *batch.Job) bool {
```
Comment (shinytang6) [inline pkg/controllers/job/job_controller_actions.go:497]: isDenpendentPodsReady  => isDependentPodsReady
DependsOnTask => task

File: pkg/controllers/job/job_controller_actions.go [hunk 10]
```diff
@@ -452,6 +468,78 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
 	return nil
 }
 
+func (cc *jobcontroller) waitDependsOnTaskMeetCondition(taskName string, podToCreateEachTask []*v1.Pod, job *batch.Job) {
+	taskIndex := jobhelpers.GetTaskIndex(taskName, job)
+	if job.Spec.Tasks[taskIndex].DependsOn != nil {
+		dependsOn := *job.Spec.Tasks[taskIndex].DependsOn
+		if len(dependsOn.Name) > 1 && dependsOn.Iteration == batch.IterationAny {
+			wait.PollInfinite(100*time.Millisecond, func() (bool, error) {
+				for _, DependsOnTask := range dependsOn.Name {
```
Comment (shinytang6) [inline pkg/controllers/job/job_controller_actions.go:477]: DependsOnTask => task

File: pkg/controllers/job/job_controller_actions.go [hunk 11]
```diff
@@ -452,6 +468,78 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+					}
+					return false, nil
+				})
+			}
+		}
+	}
+}
+
+func (cc *jobcontroller) isDenpendentPodsReady(DependsOnTask string, job *batch.Job) bool {
+	dependentPod := jobhelpers.GetPodsInTask(DependsOnTask, job)
```
Comment (shinytang6) [inline pkg/controllers/job/job_controller_actions.go:498]: dependentPod => dependentPods
And it is better to unify using dependon or dependent, not mixed

File: pkg/controllers/job/job_controller_actions.go [hunk 12]
```diff
@@ -452,6 +468,78 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+				}
+				return false
+			}
+			klog.Errorf("Failed to get pod %v/%v %v", job.Namespace, podName, err)
+			return false
+		}
+
+		if pod.Status.Phase != v1.PodRunning {
+			klog.V(4).Infof("Sequential state, pod %v/%v of dependent tasks is not running", pod.Namespace, pod.Name)
+			continue
```
Comment (shinytang6) [inline pkg/controllers/job/job_controller_actions.go:519]: what if all the pods under that task become succeeded quickly?

File: pkg/controllers/job/job_controller_actions.go [hunk 13]
```diff
@@ -452,6 +468,78 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+				_, err := cc.vcClient.BatchV1alpha1().Jobs(job.Namespace).Get(context.TODO(), job.Name, metav1.GetOptions{})
+				if err != nil {
+					if apierrors.IsNotFound(err) {
+						return true
+					}
+				}
+				return false
+			}
+			klog.Errorf("Failed to get pod %v/%v %v", job.Namespace, podName, err)
+			return false
```
Comment (shinytang6) [inline pkg/controllers/job/job_controller_actions.go:514]: l'm a little confused about this block.
1. why need to get job if this pod not found?
2. if pod is not found and job exists, do we need to return false immediately?

**Review by shinytang6 — CHANGES_REQUESTED** [review]

File: pkg/controllers/job/helpers/helpers.go [hunk 14]
```diff
@@ -34,8 +35,8 @@ const (
 	persistentVolumeClaimFmt = "%s-pvc-%s"
 )
 
-// GetTaskIndex returns task Index.
-func GetTaskIndex(pod *v1.Pod) string {
+// GetTask returns task Index.
+func GetTask(pod *v1.Pod) string {
```
Comment (shinytang6) [inline pkg/controllers/job/helpers/helpers.go:39]: And l think this func can be named more clearly, maybe `getPodIndexUnderTask`?

File: pkg/controllers/job/helpers/helpers.go (hunk 2 above)
Comment (shinytang6) [reply pkg/controllers/job/helpers/helpers.go:76]: And l think this func can be named more clearly, maybe `getTasklndexUnderJob `?

File: pkg/controllers/job/helpers/helpers.go [hunk 15]
```diff
@@ -70,3 +71,27 @@ func GenPVCName(jobName string) string {
...
+	for index, task := range job.Spec.Tasks {
+		if task.Name == taskName {
+			return index
+		}
+	}
+	return -1
+}
+
+// GetPodsInTask return names of all pods in the task.
+func GetPodsInTask(taskName string, job *batch.Job) []string {
```
Comment (shinytang6) [inline pkg/controllers/job/helpers/helpers.go:86]: And l think this func can be named more clearly, maybe `GetPodsNameUnderTask `?

**Review by shinytang6 — CHANGES_REQUESTED** [review]

File: pkg/controllers/job/job_controller_actions.go (hunk 11 above)
Comment (hwdef) [reply pkg/controllers/job/job_controller_actions.go:498]: I agree with what you said, which one do you think is better?

File: pkg/controllers/job/job_controller_actions.go (hunk 12 above)
Comment (hwdef) [reply pkg/controllers/job/job_controller_actions.go:519]: The current situation is that the following tasks will stop and will never be run.
This should be dependent on completion status, we can add a field later:

```
dependsOn:
    name: xxxxx
    status: running/complete
```


I think we can also deal with it now, if it is completed, allow it to pass.

```go
pod.Status.Phase != v1.PodRunning || pod.Status.Phase != v1.PodSucceeded
```

File: pkg/controllers/job/helpers/helpers.go (hunk 14 above)
Comment (hwdef) [reply pkg/controllers/job/helpers/helpers.go:39]: Please discuss the name of this function with @william-wang, I think both are ok

File: pkg/controllers/job/job_controller_actions.go [hunk 16]
```diff
@@ -349,33 +354,44 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+						// Failed to create Pod, waitCreationGroup a moment and then create it again
+						// This is to ensure all podsMap under the same Job created
+						// So gang-scheduling could schedule the Job successfully
+						klog.Errorf("Failed to create pod %s for Job %s, err %#v",
+							pod.Name, job.Name, err)
+						appendError(&creationErrs, fmt.Errorf("failed to create pod %s, err: %#v", pod.Name, err))
+					} else {
+						classifyAndAddUpPodBaseOnPhase(newPod, &pending, &running, &succeeded, &failed, &unknown)
+						calcPodStatus(pod, taskStatusCount)
+						klog.V(3).Infof("Created Task <%s> of Job <%s/%s>",
```
Comment (william-wang) [inline pkg/controllers/job/job_controller_actions.go:387]: This log is in for loop. it's better use log level 5.

**Comment by william-wang** [comment]
@hwdef is it possible to add e2e case for this new feature?

**Comment by hwdef** [comment]
> @hwdef is it possible to add e2e case for this new feature?

I am trying to add e2e, but due to the characteristics of this feature, I feel a little difficult.(It may also be because I am not familiar with e2e)

File: pkg/controllers/job/job_controller_actions.go [hunk 17]
```diff
@@ -355,33 +360,44 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
-				klog.V(3).Infof("Created Task <%s> of Job <%s/%s>",
-					pod.Name, job.Namespace, job.Name)
+	for taskName, podToCreateEachTask := range podToCreate {
+		if len(podToCreateEachTask) == 0 {
+			continue
+		}
+		go func(taskName string, podToCreateEachTask []*v1.Pod) {
+			taskIndex := jobhelpers.GetTasklndexUnderJob(taskName, job)
+			if job.Spec.Tasks[taskIndex].DependsOn != nil {
+				cc.waitDependsOnTaskMeetCondition(taskName, podToCreateEachTask, job)
```
Comment (william-wang) [inline pkg/controllers/job/job_controller_actions.go:376]: for each task with DependsOn, GetTaskIndexUnderJob is called twice(Line374 and Line478). It's better take taskIndex as input of waitDependsOnTaskMeetCondition.

File: pkg/controllers/job/job_controller_actions.go [hunk 18]
```diff
@@ -458,6 +474,69 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
+				for _, task := range dependsOn.Name {
+					if cc.isDependsOnPodsReady(task, job) {
+						return true, nil
+					}
+				}
+				return false, nil
+			})
+		} else {
+			for _, dependsOnTask := range dependsOn.Name {
+				wait.PollInfinite(100*time.Millisecond, func() (bool, error) {
```
Comment (william-wang) [inline pkg/controllers/job/job_controller_actions.go:492]: it's better to make 100ms configurable. At least define a const for it if no time :)

File: pkg/controllers/job/job_controller_actions.go (hunk 17 above)
Comment (hwdef) [reply pkg/controllers/job/job_controller_actions.go]: done.

File: pkg/controllers/job/job_controller_actions.go (hunk 18 above)
Comment (hwdef) [reply pkg/controllers/job/job_controller_actions.go]: I implemented configurable parameters. But I don't know if the implementation is elegant, especially the process of variable transmission. In addition, the naming of variables is also very long. If you have any comments, please let me know, thank you!

**Review by william-wang — APPROVED** [review]
/lgtm

**Pull request merged**